### PR TITLE
Fix/mounting squashfs extract

### PR DESCRIPTION
**Pull request by smoser**
**What type of PR is this?**

**Which issue does this PR fix**:

**What does this PR do / Why do we need it**:

**If an issue # is not available please add repro steps and logs showing the issue**:

**Testing done on this change**:

**Automation added to e2e**:

**Will this break upgrades or downgrades?**

**Does this PR introduce any user-facing change?**:

```release-note

```

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

**Comment by hallyn** [comment]
It seems to me that in the original code, if we do 20 ExtractSingleSquash()s, and are root in a non-init user namespace with full UID range, we will only try kernel mounting once, but with your updated code we will try it every time we call ExtractSinglesquash().  Maybe that doesn't matter, as that surely must be a very rare case.

File: pkg/squashfs/squashfs.go [hunk 1]
```diff
@@ -300,36 +289,310 @@ func squashFuse(squashFile, extractDir string) (*exec.Cmd, error) {
...
+	fdest, err := filepath.Abs(extractDir)
+	if err != nil {
 		return err
 	}
 
-	cmd, err := squashFuse(squashFile, extractDir)
-	if err == nil {
-		if err := cmd.Process.Release(); err != nil {
-			return errors.Errorf("Failed to release process %s: %v", cmd, err)
+	if policy.initialized {
```
Comment (smoser) [inline pkg/squashfs/squashfs.go:520]: this block is taken when not initialzing; it only uses policy.Extractor.

File: pkg/squashfs/squashfs.go [hunk 2]
```diff
@@ -300,36 +289,310 @@ func squashFuse(squashFile, extractDir string) (*exec.Cmd, error) {
...
+
+	if policy.Extractors == nil || len(policy.Extractors) == 0 {
+		policy.Excuses[initName] = errors.Errorf("policy had no extractors")
+		return policy.Excuses[initName]
+	}
+
+	var extractor SquashExtractor
+	var mounted bool
+	allExcuses := []string{}
+	for _, extractor = range policy.Extractors {
```
Comment (smoser) [inline pkg/squashfs/squashfs.go:548]: this block runs only when we have the initializer lock, it tries all Extractors and sets policy.Extractor on first success.

File: pkg/squashfs/squashfs.go [hunk 3]
```diff
@@ -300,36 +289,310 @@ func squashFuse(squashFile, extractDir string) (*exec.Cmd, error) {
...
+		if val == "" {
+			val = defPolicy
+		}
+		exPolInfo.policy, exPolInfo.err = NewExtractPolicy(strings.Fields(val)...)
+		if extractPolicyErr == nil {
+			for k, v := range exPolInfo.policy.Excuses {
+				log.Debugf(" squashfs extractor %s is not available: %v", k, v)
+			}
+		}
+	})
```
Comment (smoser) [inline pkg/squashfs/squashfs.go:589]: Above is how how the ExtractPolicy gets initialized.  That code only runs once (mutex.once).
After it runs, exPolInfo has filtered out the Extractors where IsAvailable == false.
So we only call IsAvailable once.

**Comment by smoser** [comment]
> It seems to me that in the original code, if we do 20 ExtractSingleSquash()s, and are root in a non-init user namespace with full UID range, we will only try kernel mounting once, but with your updated code we will try it every time we call ExtractSinglesquash(). Maybe that doesn't matter, as that surely must be a very rare case.

In the original code it  basically tries each option in order every time (kernel-mouint, squashfuse, unsquashfs).  It never evicts an option from the list due to failure.  The new code tries each option in order only once and then subsequently uses only the option that worked.

File: pkg/squashfs/squashfs.go (hunk 3 above)
Comment (hallyn) [reply pkg/squashfs/squashfs.go:589]: (I'm not sure whether these comments are replies or not)

This code only checks the AmHostRoot().  So if kernel mount fails for another reason - such as you're host root without CAP_SYS_ADMIN, or you are root in a container that happens to have the full host uid range mapped (but, not being init user ns, has CAP_SYS_ADMIN but not against the init_user_ns and so still cannot mount -t squashfs), then that kernel mount will be attempted every time.

Like I said, that's *probably* ok.

Comment (hallyn) [reply pkg/squashfs/squashfs.go:589]: Never mind - I'm wrong about this.

**Comment by smoser** [comment]
I'm happy with the state of this PR now, but I kind of expect it to fail c-i.
The reason is that builds will fail when building from a squashfs oci reference and using kernel mounts.

I'm attaching a test script that shows the failure that i'm hitting.
[build-from-squash.txt](https://github.com/project-stacker/stacker/files/12805060/build-from-squash.txt)

The failed run looks like this:

```
execute: ./stacker --debug --work-dir=tgo-data-root build --layer-type=squashfs --layer-type=tar --stacker-file=s2.yaml
stacker version v1.0.0-rc5-18-g1666822
nsexec-ing [-- /home/smoser/src/stacker/stacker --internal-userns --debug --work-dir=tgo-data-root build --layer-type=squashfs --layer-type=tar --stacker-file=s2.yaml]
stacker version v1.0.0-rc5-18-g1666822
initializing stacker recipe: s2.yaml
substituting $STACKER_ROOTFS_DIR to /home/smoser/src/stacker/tgo-data-root/roots
substituting $STACKER_STACKER_DIR to /home/smoser/src/stacker/tgo-data-root/.stacker
substituting $STACKER_OCI_DIR to /home/smoser/src/stacker/tgo-data-root/oci
substituting $STACKER_WORK_DIR to tgo-data-root
stacker build order:
0 build /home/smoser/src/stacker/s2.yaml: requires: []
building: 0 /home/smoser/src/stacker/s2.yaml
substituting $STACKER_ROOTFS_DIR to /home/smoser/src/stacker/tgo-data-root/roots
substituting $STACKER_STACKER_DIR to /home/smoser/src/stacker/tgo-data-root/.stacker
substituting $STACKER_OCI_DIR to /home/smoser/src/stacker/tgo-data-root/oci
substituting $STACKER_WORK_DIR to tgo-data-root
Dependency Order [bare2]
preparing image bare2...
overlay-dirs, possibly modified after import: []
loading oci:/home/smoser/src/stacker/tgo-data-root/oci:bare-squashfs
unpacking to /home/smoser/src/stacker/tgo-data-root/roots/bare2
Selected squashfs extractor kmount
lxc rootfs overlay arg overlayfs:/home/smoser/src/stacker/tgo-data-root/roots/sha256_b8646ebdc2a7b931b675b200d513e9a1c02c249834af67c2ed6f7df1656a95b2/overlay:/home/smoser/src/stacker/tgo-data-root/roots/sha256_12c6518b83d59b9bb87d242955a776940e68fbd6b580c4c3a92ebe074e81135c/overlay:/home/smoser/src/stacker/tgo-data-root/roots/bare2/overlay
stacker version v1.0.0-rc5-18-g1666822
stacker subcommand: [/home/smoser/src/stacker/stacker --oci-dir /home/smoser/src/stacker/tgo-data-root/oci --roots-dir /home/smoser/src/stacker/tgo-data-root/roots --stacker-dir /home/smoser/src/stacker/tgo-data-root/.stacker --storage-type overlay --debug internal-go check-aa-profile lxc-container-default-cgns]
bind mounting /home/smoser/src/stacker/tgo-data-root/.stacker/imports/bare2 into container
+ echo Building bare2
Building bare2
+ mkdir -p /layers
+ date -R
creating layer bare2 (type=tar) by converting layer bare2-squashfs (type=squashfs+verity)
error: operation not supported
get xattr list
github.com/opencontainers/umoci/oci/layer.(*tarGenerator).AddFile
        /stacker-tree/.build/gopath/pkg/mod/github.com/opencontainers/umoci@v0.4.8-0.20230920134428-7dc114a520bc/oci/layer/tar_generate.go:204
github.com/opencontainers/umoci/oci/layer.GenerateInsertLayer.func1.2
        /stacker-tree/.build/gopath/pkg/mod/github.com/opencontainers/umoci@v0.4.8-0.20230920134428-7dc114a520bc/oci/layer/generate.go:168
github.com/opencontainers/umoci/pkg/unpriv.walk
        /stacker-tree/.build/gopath/pkg/mod/github.com/opencontainers/umoci@v0.4.8-0.20230920134428-7dc114a520bc/pkg/unpriv/unpriv.go:537
github.com/opencontainers/umoci/oci/layer.GenerateInsertLayer.func1.Walk.func3
        /stacker-tree/.build/gopath/pkg/mod/github.com/opencontainers/umoci@v0.4.8-0.20230920134428-7dc114a520bc/pkg/unpriv/unpriv.go:576
github.com/opencontainers/umoci/pkg/unpriv.Wrap
        /stacker-tree/.build/gopath/pkg/mod/github.com/opencontainers/umoci@v0.4.8-0.20230920134428-7dc114a520bc/pkg/unpriv/unpriv.go:75
github.com/opencontainers/umoci/pkg/unpriv.Walk
        /stacker-tree/.build/gopath/pkg/mod/github.com/opencontainers/umoci@v0.4.8-0.20230920134428-7dc114a520bc/pkg/unpriv/unpriv.go:571
github.com/opencontainers/umoci/oci/layer.GenerateInsertLayer.func1
        /stacker-tree/.build/gopath/pkg/mod/github.com/opencontainers/umoci@v0.4.8-0.20230920134428-7dc114a520bc/oci/layer/generate.go:153
runtime.goexit
        /usr/lib/go/src/runtime/asm_amd64.s:1650
unpriv.walk
github.com/opencontainers/umoci/oci/layer.GenerateInsertLayer.func1.Walk.func3
        /stacker-tree/.build/gopath/pkg/mod/github.com/opencontainers/umoci@v0.4.8-0.20230920134428-7dc114a520bc/pkg/unpriv/unpriv.go:581
github.com/opencontainers/umoci/pkg/unpriv.Wrap
        /stacker-tree/.build/gopath/pkg/mod/github.com/opencontainers/umoci@v0.4.8-0.20230920134428-7dc114a520bc/pkg/unpriv/unpriv.go:75
github.com/opencontainers/umoci/pkg/unpriv.Walk
        /stacker-tree/.build/gopath/pkg/mod/github.com/opencontainers/umoci@v0.4.8-0.20230920134428-7dc114a520bc/pkg/unpriv/unpriv.go:571
github.com/opencontainers/umoci/oci/layer.GenerateInsertLayer.func1
        /stacker-tree/.build/gopath/pkg/mod/github.com/opencontainers/umoci@v0.4.8-0.20230920134428-7dc114a520bc/oci/layer/generate.go:153
runtime.goexit
        /usr/lib/go/src/runtime/asm_amd64.s:1650
generate insert layer
github.com/opencontainers/umoci/oci/layer.GenerateInsertLayer.func1.1
        /stacker-tree/.build/gopath/pkg/mod/github.com/opencontainers/umoci@v0.4.8-0.20230920134428-7dc114a520bc/oci/layer/generate.go:140
github.com/opencontainers/umoci/oci/layer.GenerateInsertLayer.func1
        /stacker-tree/.build/gopath/pkg/mod/github.com/opencontainers/umoci@v0.4.8-0.20230920134428-7dc114a520bc/oci/layer/generate.go:153
runtime.goexit
        /usr/lib/go/src/runtime/asm_amd64.s:1650
copy to temporary blob
github.com/opencontainers/umoci/oci/cas/dir.(*dirEngine).PutBlob
        /stacker-tree/.build/gopath/pkg/mod/github.com/opencontainers/umoci@v0.4.8-0.20230920134428-7dc114a520bc/oci/cas/dir/dir.go:173
stackerbuild.io/stacker/pkg/overlay.ociPutBlob
        /stacker-tree/pkg/overlay/pack.go:277
stackerbuild.io/stacker/pkg/overlay.ConvertAndOutput
        /stacker-tree/pkg/overlay/pack.go:127
stackerbuild.io/stacker/pkg/overlay.(*overlay).initializeBasesInOutput
        /stacker-tree/pkg/overlay/pack.go:212
stackerbuild.io/stacker/pkg/overlay.(*overlay).Repack
        /stacker-tree/pkg/overlay/pack.go:242
stackerbuild.io/stacker/pkg/stacker.(*Builder).build
        /stacker-tree/pkg/stacker/build.go:512
stackerbuild.io/stacker/pkg/stacker.(*Builder).BuildMultiple
        /stacker-tree/pkg/stacker/build.go:610
main.doBuild
        /stacker-tree/cmd/stacker/build.go:118
github.com/urfave/cli/v2.(*Command).Run
        /stacker-tree/.build/gopath/pkg/mod/github.com/urfave/cli/v2@v2.25.0/command.go:273
github.com/urfave/cli/v2.(*Command).Run
        /stacker-tree/.build/gopath/pkg/mod/github.com/urfave/cli/v2@v2.25.0/command.go:266
github.com/urfave/cli/v2.(*App).RunContext
        /stacker-tree/.build/gopath/pkg/mod/github.com/urfave/cli/v2@v2.25.0/app.go:332
github.com/urfave/cli/v2.(*App).Run
        /stacker-tree/.build/gopath/pkg/mod/github.com/urfave/cli/v2@v2.25.0/app.go:309
main.main
        /stacker-tree/cmd/stacker/main.go:328
runtime.main
        /usr/lib/go/src/runtime/proc.go:267
runtime.goexit
        /usr/lib/go/src/runtime/asm_amd64.s:1650
error: exit status 1
stackerbuild.io/stacker/pkg/container.MaybeRunInNamespace
        /stacker-tree/pkg/container/userns.go:102
main.main.func3
        /stacker-tree/cmd/stacker/main.go:323
github.com/urfave/cli/v2.(*Command).Run
        /stacker-tree/.build/gopath/pkg/mod/github.com/urfave/cli/v2@v2.25.0/command.go:213
github.com/urfave/cli/v2.(*App).RunContext
        /stacker-tree/.build/gopath/pkg/mod/github.com/urfave/cli/v2@v2.25.0/app.go:332
github.com/urfave/cli/v2.(*App).Run
        /stacker-tree/.build/gopath/pkg/mod/github.com/urfave/cli/v2@v2.25.0/app.go:309
main.main
        /stacker-tree/cmd/stacker/main.go:328
runtime.main
        /usr/lib/go/src/runtime/proc.go:267
runtime.goexit
        /usr/lib/go/src/runtime/asm_amd64.s:1650
failed build s2.yaml
```

**Comment by hallyn** [comment]
```
error: operation not supported
get xattr list
```
this looks like another place where caller needs to accept that squashsfs/overlay can return EOPNOTSUPP for some files while other files support it

**Comment by hallyn** [comment]
> ```
> error: operation not supported
> get xattr list
> ```
> 
> this looks like another place where caller needs to accept that squashsfs/overlay can return EOPNOTSUPP for some files while other files support it

It is fixed on my system by building with https://github.com/opencontainers/umoci/pull/506

**Review by hallyn — APPROVED** [review]

**Comment by smoser** [comment]
Some status on where we are right now on this.
a. it passed c-i in the last run (except for the code coverage)
b. the fact that we are now using kernel squashfs mounts means my provided test case above fails because squashfs does not implement xattrs on a dir. this will require us to pull in serge's linked fix (https://github.com/opencontainers/umoci/pull/506).
c. there is still one issue that needs fixing; in the event that we built a layer its roots/<name>/overlay dir might have been populated by generateLayer.  In taht case we are currently mounting the squashfs derived image over the top of that overlay dir and we do not have to do that.  the general idea in stacker overlay storage is "if 'overlay' exists, then it is populated". that is somewhat sane in that it makes use of the "mkdir" semaphore.
d. i dont' love the scattered locking and i'm not sure what to do about it. i'm considering changing how that is done.  i really would like for "extraction" of a layers to be able to be done in parallel.

**Comment by hallyn** [comment]
We could keep inside the stacker binary an empty squashfs; every time stacker starts, we try to mount it from and to a tmpdir, and use that to inform whether to use kernel mounting...

It may seem wasteful, but apart from not having to do the Once() for the first kernel mount of an image, it also prevents us from misinterpreting an attempted mount of a bad squashfs blob as insufficient privilege.

**Comment by smoser** [comment]
> We could keep inside the stacker binary an empty squashfs; every time stacker starts, we try to mount it from and to a tmpdir, and use that to inform whether to use kernel mounting...
> 
> It may seem wasteful, but apart from not having to do the Once() for the first kernel mount of an image, it also prevents us from misinterpreting an attempted mount of a bad squashfs blob as insufficient privilege.

I kind of like it.
```
$ echo "hi mom" > file.txt
$ mksquashfs file.txt out.squashfs
$ ls -l out.squashfs
-rw-r--r-- 1 smoser smoser 4096 Oct  5 16:09 out.squashfs
```

4096 bytes isn't too wasteful.

**Comment by smoser** [comment]
@hallyn since you reviewed/approved I did a few things, which can be seen [here](https://github.com/project-stacker/stacker/compare/39984eaed8cd0b4018c5d21e12c425bde1e2467a..b8fba13943162461533eada7faffd9c043e6628f).

1. added some more debug (the top commit here)
2. changed the Mount interface:
```diff
- Mount(source, dest string) (bool, error)
+ Mount(source, dest string) error
```
the 'bool' was intended to be "is mounted", but in all cases if err was != nil then mounted was true.

3. avoid calling unpackOne for duplicate digests within an image (an image could have multiple layers with the  same digest. if it did, calling unpackOne on that would just likely expose races as it is right now).

4. avoid a mount or extract in unpackOne if the target exists and is populated; also add a comment here.

**Pull request merged**